mlir-enumerate/mlir-enumerate.cpp

```diff
@@ -247,6 +247,73 @@ LogicalResult addOperation(GeneratorInfo &info) {
   return success();
 }
 
+/// Check if the given OperationOp can use specified type as a valid return
+/// type.
+bool isValidReturnType(GeneratorInfo &info, irdl::OperationOp op,
+                       mlir::Type type) {
+  auto builder = info.builder;
+  auto ctx = builder.getContext();
+
+  auto constraintOp = op.getOp<irdl::ConstraintVarsOp>();
+  auto resultDefs = op.getOp<ResultsOp>();
+
+  SmallVector<std::pair<StringRef, std::unique_ptr<irdl::TypeConstraint>>>
+      namedConstraintVars = {};
+  SmallVector<std::unique_ptr<irdl::TypeConstraint>> varConstraints;
+  SmallVector<Type> vars;
+
+  // For each constraint variable, we try to assign the constraint with the
+  // provided type. Otherwise use one random from all satisfied types.
+  if (constraintOp) {
+    for (auto namedConstraintAttr : constraintOp->getParams()) {
+      auto namedConstraint =
+          namedConstraintAttr.cast<NamedTypeConstraintAttr>();
+      auto constraint =
+          namedConstraint.getConstraint()
+              .cast<TypeConstraintAttrInterface>()
+              .getTypeConstraint(info.irdlContext, namedConstraintVars);
+      // TODO(fehr) Currently a hack, will be fixed later once I update
+      // the IRDL API.
+      auto constraint2 =
+          namedConstraint.getConstraint()
+              .cast<TypeConstraintAttrInterface>()
+              .getTypeConstraint(info.irdlContext, namedConstraintVars);
+
+      if (constraint.get()
```

Comment:
You do not actually need to assign any type on the constraints here.
This should be done automatically when you check that the result types are satisfying the constraints.
So when you check something like `Vector<T>`  with the type `Vector<i32>` for instance, `T` will be set to `i32` in the `verifyType` function.

So for here, you can actually just check the result constraints, without dealing manually with the constraint variables.
You still though have to fill `namedConstraintVars`, `varConstraints`, and `vars`. But you should just set each `Type` to `Type()` in `vars` if I recall correctly.